pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,711 @@
+
+	return res, err
+}
+
+type Transaction struct {
+	/* ovs */
+	dbname     string
+	operations []libovsdb.Operation
+	response   libovsdb.TransactResponse
+
+	/* etcd */
+	etcdEndpoints []string
+	etcdCli       *clientv3.Client
+	etcdCtx       context.Context
+	etcdIf        []clientv3.Cmp
+	etcdThen      []clientv3.Op
+	etcdElse      []clientv3.Op
+	etcdCache     map[string]map[string]map[string]interface{}
+}
+
+func NewTransaction(dbname string, operations []libovsdb.Operation, endpoints []string) *Transaction {
+	txn := new(Transaction)
+	txn.dbname = dbname
+	txn.operations = operations
+	txn.response.Result = make([]libovsdb.OperationResult, len(operations))
+	txn.etcdCtx = context.TODO()
+	txn.etcdEndpoints = endpoints
+	return txn
+}
+
+func (txn *Transaction) Open() error {
+	cli, err := clientv3.New(clientv3.Config{Endpoints: txn.etcdEndpoints})
+	txn.etcdCli = cli
+	return err
+}
+
+func (txn *Transaction) Close() {
+	txn.etcdCli.Close()
+}
+
+type ovsOpCallback func(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error
+
+var ovsOpCallbackMap = map[string][]ovsOpCallback{
+	OP_INSERT:  {preInsert, doInsert},
+	OP_SELECT:  {preSelect, doSelect},
+	OP_UPDATE:  {preUpdate, doUpdate},
+	OP_MUTATE:  {preMutate, doMutate},
+	OP_DELETE:  {preDelete, doDelete},
+	OP_WAIT:    {preWait, doWait},
+	OP_COMMIT:  {preCommit, doCommit},
+	OP_ABORT:   {preAbort, doAbort},
+	OP_COMMENT: {preComment, doComment},
+	OP_ASSERT:  {preAssert, doAssert},
+}
+
+func (txn *Transaction) Commit() error {
+	var err error
+
+	/* fetch needed data from database needed to perform the operation */
+	for i, ovsOp := range txn.operations {
+		if err = ovsOpCallbackMap[ovsOp.Op][0](txn, &ovsOp, &txn.response.Result[i]); err != nil {
+			txn.response.Error = err.Error()
+			return err
+		}
+	}
+	_, err = txn.etcdCommit()
+	if err != nil {
+		txn.response.Error = err.Error()
+		return err
+	}
+
+	/* commit actual transactional changes to database */
+	for i, ovsOp := range txn.operations {
+		ovsOpCallbackMap[ovsOp.Op][1](txn, &ovsOp, &txn.response.Result[i])
+		txn.response.Error = err.Error()
+		return err
+	}
+	_, err = txn.etcdCommit()
+	if err != nil {
+		txn.response.Error = err.Error()
+		return err
+	}
+
+	return nil
+}
+
+const DELIM = "/"
+const UUID = "_uuid"
+
+func makePrefix(dbname, table string) string {
+	return dbname + DELIM + table
+}
+
+func makeKey(dbname, table, uuid string) string {
+	return dbname + DELIM + table + DELIM + uuid
+}
+
+func makeValue(row map[string]interface{}) (string, error) {
+	b, err := json.Marshal(row)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func setRowUUID(row map[string]interface{}, uuid string) {
+	row[UUID] = uuid
+}
+
+const (
+	FN_LT = "<"
+	FN_LE = "<="
+	FN_EQ = "=="
+	FN_NE = "!="
+	FN_GE = ">="
+	FN_GT = ">"
+	FN_IN = "includes"
+	FN_EX = "excludes"
+)
+
+func isSupportedFunc(function string) error {
+	if function == FN_EQ || function == FN_NE {
```

Comment:
should we check only these 2 ?

Reply:
agree, I have removed this and now we check everything according to spec